pkg/api/tag_test.go

```diff
@@ -0,0 +1,237 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package api_test
+
+import (
+	"bytes"
+	"net/http"
+	"strconv"
+	"testing"
+
+	"github.com/ethersphere/bee/pkg/api"
+	"github.com/ethersphere/bee/pkg/jsonhttp"
+	"github.com/ethersphere/bee/pkg/jsonhttp/jsonhttptest"
+	mp "github.com/ethersphere/bee/pkg/pusher/mock"
+	"github.com/ethersphere/bee/pkg/storage/mock"
+	"github.com/ethersphere/bee/pkg/storage/mock/validator"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/ethersphere/bee/pkg/tags"
+)
+
+func TestTags(t *testing.T) {
+	var (
+		resource             = func(addr swarm.Address) string { return "/bzz-chunk/" + addr.String() }
+		tagResourceAddress   = func(addr swarm.Address) string { return "/bzz-tag/addr/" + addr.String() }
+		tagResourceUidCreate = func(uid string) string { return "/bzz-tag/name/" + uid }
+		tagResourceUUid      = func(uuid uint64) string { return "/bzz-tag/uuid/" + strconv.FormatUint(uuid, 10) }
+		validHash            = swarm.MustParseHexAddress("aabbcc")
+		validContent         = []byte("bbaatt")
+		mockValidator        = validator.NewMockValidator(validHash, validContent)
+		tag                  = tags.NewTags()
+		mockValidatingStorer = mock.NewValidatingStorer(mockValidator, tag)
+		mockPusher           = mp.NewMockPusher(tag)
+		client               = newTestServer(t, testServerOptions{
+			Storer: mockValidatingStorer,
+			Tags:   tag,
+		})
+	)
+
+	t.Run("send-invalid-tag-id", func(t *testing.T) {
+		sentHheaders := make(http.Header)
+		sentHheaders.Set(api.TagHeaderUid, "file.jpg") // the value should be uint32
+		_ = jsonhttptest.ResponseDirectSendHeadersAndReceiveHeaders(t, client, http.MethodPost, resource(validHash), bytes.NewReader(validContent), http.StatusBadRequest, jsonhttp.StatusResponse{
+			Message: "invalid taguid",
+			Code:    http.StatusBadRequest,
+		}, sentHheaders)
+	})
+
+	t.Run("uid-header-in-return-for-empty-tag", func(t *testing.T) {
+		rcvdHeaders := jsonhttptest.ResponseDirectSendHeadersAndReceiveHeaders(t, client, http.MethodPost, resource(validHash), bytes.NewReader(validContent), http.StatusOK, jsonhttp.StatusResponse{
+			Message: http.StatusText(http.StatusOK),
+			Code:    http.StatusOK,
+		}, nil)
+
+		isTagFoundInResponse(t, rcvdHeaders, nil)
+	})
+
+	t.Run("get-tag-and-use-it-to-upload-chunk", func(t *testing.T) {
+		// Get a tag using API
+		ta := &tags.Tag{}
```

Comment:
To be completely correct, now, the response is api.TagResponse (can be exported in export_test.go). This type should be unmarshalled to, while leaving tags.Tag independent of API tests. This applies to other subtests in this package.